src/mpi/ccl/cclcomm.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpiimpl.h"
+
+#define CUDA_ERR_CHECK(ret) if (unlikely((ret) != cudaSuccess)) goto fn_fail
+
+/*
+ * CCLcomm functions, currently tied to NCCL
+ */
+
+int MPIR_CCLcomm_init(MPIR_Comm * comm, int rank)
+{
+    int mpi_errno = MPI_SUCCESS;
+    cudaError_t ret;
+    int comm_size = comm->local_size;
+
+    MPIR_CCLcomm *cclcomm;
+    cclcomm = MPL_malloc(sizeof(MPIR_CCLcomm), MPL_MEM_OTHER);
+    MPIR_ERR_CHKANDJUMP(!cclcomm, mpi_errno, MPI_ERR_OTHER, "**nomem");
+
+    if(rank == 0) {
+        ncclGetUniqueId(&(cclcomm->id));
+    }
+
+    mpi_errno = MPIR_Bcast_impl(&(cclcomm->id), sizeof(cclcomm->id), MPI_INT, 0, comm,
+                                    MPIR_ERR_NONE);
+    MPIR_ERR_CHECK(mpi_errno);
+    
+    ret = cudaSetDevice(0); //TODO: Fix this for multi-GPU nodes
+    CUDA_ERR_CHECK(ret);
+    ret = cudaStreamCreate(&(cclcomm->stream));
+    CUDA_ERR_CHECK(ret);
+    ret = ncclCommInitRank(&(cclcomm->ncclcomm), comm_size, cclcomm->id, rank);
```

Comment:
Does NCCL work with duplicate/shared devices? Does it work across multiple nodes?

Reply:
I guess the question is whether nccl's internode support has advantage. I think it is simpler if we limit MPICH's usage of nccl to intranode, and rely/implement heirarchical algorithm for internode collectives.

Reply:
I plan to support intranode and internode using CCLs. NCCL has performance advantages for internode, and other CCLs do too.